### ISSUE TITLE
[engine] Security expression `permitAll()` not working

### ISSUE THREAD

**Issue by jdrossl** [craftercms/craftercms#4392]
## Describe the bug
The security expression `permitAll()` is not working as described in the our docs https://docs.craftercms.org/en/3.1/site-administrators/engine/engine-site-security-guide.html

## To Reproduce
1. Add a url restriction like `/** -> permitAll()`
2. Go to any URL and see it fail

## Expected behavior
URLs with `permitAll()` should always be public

## Logs
```
[DEBUG] 2020-11-09T11:09:57,543 [http-nio-8080-exec-8] [test] [intercept.FilterSecurityInterceptor] | Secure object: FilterInvocation: URL: /; Attributes: [permitAll()] 
[DEBUG] 2020-11-09T11:10:04,826 [http-nio-8080-exec-8] [test] [access.ExceptionTranslationFilter] | Authentication exception occurred; redirecting to authentication entry point 
org.springframework.security.authentication.AuthenticationCredentialsNotFoundException: An Authentication object was not found in the SecurityContext
```

### Version
3.1.x & 3.2

**Comment by jdrossl** [comment]
@yacdaniel please help verify

**Comment by yacdaniel** [comment]
Verified fixed in `3.1.12`

**Comment by yacdaniel** [comment]
Verified fixed in `4.0.0`
